pvlib/pvsystem.py

```diff
@@ -66,6 +69,57 @@ def _combine_localized_attributes(pvsystem=None, location=None, **kwargs):
     return new_kwargs
 
 
+def return_singleton_value(func):
+    """Decorator that returns the value contained when `func`
+    returns a singleton list or the list if it has more than one
+    element.
+    """
+    @functools.wraps(func)
+    def f(*args, **kwargs):
+        return_list = kwargs.pop('return_list', False)
+        x = func(*args, **kwargs)
+        if len(x) == 1 and not return_list:
+            return x[0]
+        return x
+    return f
+
+
+def validate_against_arrays(*list_params):
+    """Decorator that validates the value passed to each parameter in
+    `list_params` against the number of Arrays in the PVSystem.
+
+    If the value passed for each parameter in `list_params` is not a list,
+    then it is transformed into a singleton list before validation. This
+    means existing code that assumes a PVSystem has only one array will
+    continue to function with no changes.
+
+    Implicitly applies the `@singleton_as_scalar` decorator as well.
+
+    Parameters
+    ----------
+    list_params : iterable
+        names of the parameters that should be lists with the same number
+        of elements as there are Arrays in the PVSystem instance.
+    """
+    def list_or_scalar(func):
+        @return_singleton_value
+        @functools.wraps(func)
+        def f(ref, *args, **kwargs):
+            sig = inspect.signature(func)
```

Comment:
This is a lot of complexity for pvlib. It's in the object layer so I'm not entirely opposed, but I think it's worth discussing if we really want to commit to maintaining this. I've failed to develop my decorator skills despite repeated attempts.

Reply:
The use on introspection makes me slightly uneasy as well. I think there is a trade off between replicating a lot of list-validation logic in each function and the complexity of introspection. Without the decorator I think we would have code that looks like this in each function.

```python
class PVSystem:
    # ...
    def _validate_array_input(self, x):
        if not isinstance(x, list):
            x = [x]
        if len(x) != len(self._arrays):
            raise ValueError(...)
        return x

    def sapm_effective_irradiance(self, poa_direct, poa_diffuse,
                                  airmass_absolute, aoi,
                                  reference_irradiance=1000):
        poa_direct = self._validate_array_input(poa_direct)
        poa_diffuse = self._validate_array_input(poa_diffuse)
        aoi = self._validate_array_input(aoi)
        return [ ... ]
```
 
I actually don't mind this, and it is certainly more understandable than the decorator approach. I may have tried to be too clever for my own good.

Reply:
One more thought: with the non-decorator approach, I think backtraces from the `ValueError` will be easier for users to understand.

Reply:
Also a good idea. I don't have strong feelings about the right approach at this point. I don't think you were being too clever, it's just a question of whether it's a good fit for this part of this library. I'm curious what other users and maintainers think.

Reply:
@cwhanse Do you have any thoughts on this? I'm fairly conflicted, but am leaning towards changing to the explicit validation approach in the comment above. I think it is difficult to justify the added complexity of introspection, and the explicit validataion is more clear in many ways.